File: data_structures/trees/binary_search_tree/typescript/BinarySearchTreeNode.ts
import { BaseBinaryTreeNodeInterface } from '@ds/trees/types';

export type BinarySearchTreeNodeLeafType<T> = BinarySearchTreeNode<T> | null;

export default class BinarySearchTreeNode<T> implements BaseBinaryTreeNodeInterface<T> {
    left: BinarySearchTreeNodeLeafType<T> = null;
    right: BinarySearchTreeNodeLeafType<T> = null;
    parent: BinarySearchTreeNodeLeafType<T> = null;
    val: T;

    constructor(val: T = null, parent: BinarySearchTreeNode<T> = null) {
        this.val = val;
        this.parent = parent;
    }
}

File: data_structures/trees/binary_search_tree/python/binary_search_tree_node.py
from typing import Generic, TypeVar, Union

T = TypeVar('T')
BinarySearchTreeNodeType = Union['BinarySearchTreeNode[T]', None]


class BinarySearchTreeNode(Generic[T]):
    left: BinarySearchTreeNodeType = None
    right: BinarySearchTreeNodeType = None
    parent: BinarySearchTreeNodeType = None
    val: T = None

    def __init__(self, val: Union[T, None] = None, parent: BinarySearchTreeNodeType = None):
        self.val = val
        self.parent = parent
